msprime/simulations.py

```diff
@@ -1471,6 +1489,178 @@ def print_history(self, output=sys.stdout):
+            N_t[j] = N
+        return N_t
+
+    def coalescence_rate_trajectory(self, steps, num_samples, min_pop_size=1,
+                                    double_step_validatation=True):
+        """
+        This function will calculate the mean coalescence rates between the
+        lineages of the sample configuration provided in `num_samples`, at
+        each generation time listed by `steps`, in this demographic model.
+        The coalescence rate at time t in the past is the average rate
+        of coalescence of as-yet-uncoalesed lineages, computed as follows:
+        let p(t) be the probability that the lineages of a randomly
+        chosen pair of samples has not yet coalesced by time t,
+        let p(z,t) be the probability that the lineages of a randomly
+        chosen pair of samples has not yet coalesced by time t *and* are
+        both in population z, and let N(z,t) be the diploid effective
+        population size of population z at time t. Then the mean coalescence
+        rate at time t is r(t) = (sum_z p(z,t) / (2 * N(z,t) ) / p(t).
+        The computation is done approximating population size trajetories
+        with piecewise constant trajectories, in steps of size dt = end / num_steps.
+        For this to be accurate, dt must be small enough so that (a) short epochs
+        (e.g., bottlenecks) are not missed, and (b) populations do not
+        change in size too much over that time, if they are growing or shrinking.
+        If in doubt, check that results are simliar after doubling `num_steps`.
+
+        :param list steps: The respective time steps (in generations) for which the
+            coalescence rate, as well as probablility that a randomly pair of lineages
+            has not yet coalesced, is computed and returned. The list given should be
+            sorted in ascending order.
+
+        :param list num_samples: is a list of the same length as the number
+            of populations, so that `num_samples[j]` is the number of sampled
+            chromosomes in subpopulation `j`.
+
+        :param int min_pop_size: is the smallest allowed population size during
+            computation of coalescent rates (i.e., coalescence rates are actually
+            1 / 2 * max(min_pop_size, N(z,t))). Spurious very small population sizes
+            can occurIn models where populations grow exponentially but are unused
+            before some time in the past, and lead to floating point error.
+            This should be set to a value smaller than the smallest
+            desired population size in the model.
+
+        :return: Returns a tuple where the first value is an array whose jth
+            element is the coalescence rate at the jth time point
+            (denoted r(t[j]) above), and the second is the probablility that a
+            randomly pair of lineages has not yet coalesced (denoted p(t[j]) above).
+
+        :rtype: class:'tuple'
+        """
```

Comment:
We should probably do more user parameter input checking here, so we can throw useful errors/warnings. 
check that:
1: num_samples has len equal to num_pops
2: The number of samples doesn't exceed the number being sampled for the demographic model (maybe not?)
3: min_pop_size is sufficient? (maybe not?)
4: the steps are sequential?
5: Other requirements?

Reply:
If I knew which specific `msprime` errors/ warnings I should throw I could get this done real quick.

Reply:
1: num_samples has len equal to num_pops

`raise ValueError("num_samples must have the same length as the number of populations")`

2: The number of samples doesn't exceed the number being sampled for the demographic model (maybe not?)

No need for this check.

3: min_pop_size is sufficient? (maybe not?)

I think not: we don't have a good way of doing this beforehand.

4: the steps are sequential?

`raise ValueError("steps must be a sequence of increasing times.")`

5: Other requirements?

No zero-size steps.

We need documentation for `double_step_validation` also (or just *always* do it: this is my vote actually).

Reply:
Sounds good

> We need documentation for double_step_validation also (or just always do it: this is my vote actually).

Happy to do this, my two cents: I think that If I were to, for some reason put this in a pipeline where I could guarantee my step sizes correct (by testing the domain once) and wanted to do this in a loop, I would want the option to cut the time down.

Reply:
Errors looks good to me --- we need some tests to provoke them though. See elsewhere for examples of testing error inputs using ``self.assertRaises``